client/app/components/dynamic-form/DynamicForm.jsx

```diff
@@ -0,0 +1,228 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Form, Input, InputNumber, Checkbox, Button, Upload, Icon } from 'antd';
+import { react2angular } from 'react2angular';
+import { toastr } from '@/services/toastr';
+import { Field, Action, AntdForm } from '../proptypes';
+import helper from './dynamicFormHelper';
+
+export class DynamicForm extends React.Component {
+  static propTypes = {
+    fields: PropTypes.arrayOf(Field),
+    actions: PropTypes.arrayOf(Action),
+    onSubmit: PropTypes.func,
+    form: AntdForm.isRequired,
+  };
+
+  static defaultProps = {
+    fields: [],
+    actions: [],
+    onSubmit: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      isSubmitting: false,
+      inProgressActions: [],
+    };
+
+    this.actionCallbacks = this.props.actions.reduce((acc, cur) => ({
+      ...acc,
+      [cur.name]: cur.callback,
+    }), []);
```

Comment:
Nice trick :)
Btw, the `[]` argument should be `{}` or even better `null`.

Reply:
Good catch, for some reason I thought with `null` this wouldn't work...